File: clients/transmission.py
"""Transmission client integration."""

from typing import Union
import urllib.parse
import transmission_rpc
from transmission_rpc.error import (
    TransmissionAuthError,
    TransmissionConnectError,
    TransmissionTimeoutError,
)

from helpers.config import SpeedrrConfig, ClientConfig
from helpers.log_loader import logger
from helpers.bit_convert import bit_conv


class TransmissionClient:
    """Transmission client wrapper."""

    def __init__(self, config: SpeedrrConfig, config_client: ClientConfig) -> None:
        self._client_config = config_client
        self._config = config

        # Gets hostname, port, and path from url and checks if values are sensible
        u = urllib.parse.urlparse(config_client.url)

        protocol = u.scheme
        if protocol == "http":
            default_port = 80
        elif protocol == "https":
            default_port = 443
        else:
            raise ValueError(f"<trans|{self._client_config.url}> Unknown url scheme {u.scheme}")

        if u.hostname is None:
            raise ValueError(f"<trans|{self._client_config.url}> Missing hostname")

        logger.debug(
            "<trans|%s> Connecting to Transmission at %s",
            self._client_config.url,
            config_client.url
        )

        try:
            self._client = transmission_rpc.Client(
                protocol=protocol,
                username=config_client.username,
                password=config_client.password,
                host=u.hostname,
                port=u.port or default_port,
                path=u.path or "/transmission/rpc",
            )
        except TransmissionTimeoutError as err:
            raise RuntimeError(
                f"<trans|{self._client_config.url}> Connection to Transmission timed out"
            ) from err
        except TransmissionAuthError as err:
            raise RuntimeError(
                f"<trans|{self._client_config.url}> Failed to login to Transmission, "
                "check your credentials"
            ) from err
        except TransmissionConnectError as err:
            raise RuntimeError(
                f"<trans|{self._client_config.url}> Failed to connect to Transmission, "
                "check your url"
            ) from err

        logger.debug("<trans|%s> Connected to Transmission", self._client_config.url)

    @property
    def client_config(self) -> ClientConfig:
        """Return client configuration."""
        return self._client_config

    def get_active_torrent_count(self) -> int:
        """Get the number of torrents that are currently downloading or uploading."""
        logger.debug("<trans|%s> Getting active torrent count", self._client_config.url)

        session_stats = self._client.session_stats()
        return session_stats.active_torrent_count

    def set_upload_speed(self, speed: Union[int, float]) -> None:
        """Set the upload speed limit for the client, in config units."""
        if speed == float('inf'):
            logger.debug("<trans|%s> Setting upload speed to unlimited", self._client_config.url)
            self._client.set_session(speed_limit_up_enabled=False)
        else:
            logger.debug(
                "<trans|%s> Setting upload speed to %s%s",
                self._client_config.url, speed, self._config.units
            )
            speed_limit_up = max(1, int(bit_conv(speed, self._config.units, 'KB')))
            self._client.set_session(
                speed_limit_up_enabled=True, speed_limit_up=speed_limit_up
            )

    def set_download_speed(self, speed: Union[int, float]) -> None:
        """Set the download speed limit for the client, in config units."""
        if speed == float('inf'):
            logger.debug(
                "<trans|%s> Setting download speed to unlimited", self._client_config.url
            )
            self._client.set_session(speed_limit_down_enabled=False)
        else:
            logger.debug(
                "<trans|%s> Setting dowload speed to %s%s",
                self._client_config.url, speed, self._config.units
            )
            speed_limit_down = max(1, int(bit_conv(speed, self._config.units, "KB")))
            self._client.set_session(
                speed_limit_down_enabled=True, speed_limit_down=speed_limit_down
            )

File: client/transmission.go
package client

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"sync"
	"time"

	"github.com/khw315/speedrr/config"
	"github.com/khw315/speedrr/logger"
	"github.com/khw315/speedrr/units"
)

type TransmissionClient struct {
	mu           sync.Mutex
	appConfig    *config.SpeedrrConfig
	clientConfig config.ClientConfig
	httpClient   *http.Client
	rpcURL       string
	sessionID    string
}

type rpcRequest struct {
	Method    string      `json:"method"`
	Arguments interface{} `json:"arguments,omitempty"`
}

type rpcResponse struct {
	Result    string                 `json:"result"`
	Arguments map[string]interface{} `json:"arguments"`
}

func NewTransmissionClient(appConfig *config.SpeedrrConfig, clientConfig config.ClientConfig) (*TransmissionClient, error) {
	u, err := url.Parse(clientConfig.URL)
	if err != nil {
		return nil, fmt.Errorf("<trans|%s> invalid URL: %w", clientConfig.URL, err)
	}

	rpcPath := u.Path
	if rpcPath == "" || rpcPath == "/" {
		rpcPath = "/transmission/rpc"
	}
	u.Path = rpcPath
	rpcURL := u.String()

	transport := &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: !clientConfig.HTTPSVerify,
		},
	}

	httpClient := &http.Client{
		Transport: transport,
		Timeout:   15 * time.Second,
	}

	c := &TransmissionClient{
		appConfig:    appConfig,
		clientConfig: clientConfig,
		httpClient:   httpClient,
		rpcURL:       rpcURL,
	}

	logger.Debug("<trans|%s> Connecting to Transmission at %s", clientConfig.URL, clientConfig.URL)
	// Test connection via session-stats call
	_, err = c.GetActiveTorrentCount(context.Background())
	if err != nil {
		return nil, fmt.Errorf("<trans|%s> Failed to connect to Transmission: %w", clientConfig.URL, err)
	}
	logger.Debug("<trans|%s> Connected to Transmission", clientConfig.URL)

	return c, nil
}

func (c *TransmissionClient) Config() config.ClientConfig {
	return c.clientConfig
}

func (c *TransmissionClient) doRPC(ctx context.Context, method string, args interface{}) (*rpcResponse, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	reqBody := rpcRequest{
		Method:    method,
		Arguments: args,
	}
	data, err := json.Marshal(reqBody)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.rpcURL, bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.clientConfig.Username != "" || c.clientConfig.Password != "" {
		req.SetBasicAuth(c.clientConfig.Username, c.clientConfig.Password)
	}
	if c.sessionID != "" {
		req.Header.Set("X-Transmission-Session-Id", c.sessionID)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Handle Transmission 409 Conflict (CSRF Session ID requirement)
	if resp.StatusCode == http.StatusConflict {
		newSessionID := resp.Header.Get("X-Transmission-Session-Id")
		if newSessionID != "" {
			c.sessionID = newSessionID
			// Retry request once with new session ID
			reqRetry, err := http.NewRequestWithContext(ctx, http.MethodPost, c.rpcURL, bytes.NewReader(data))
			if err != nil {
				return nil, err
			}
			reqRetry.Header.Set("Content-Type", "application/json")
			if c.clientConfig.Username != "" || c.clientConfig.Password != "" {
				reqRetry.SetBasicAuth(c.clientConfig.Username, c.clientConfig.Password)
			}
			reqRetry.Header.Set("X-Transmission-Session-Id", c.sessionID)

			respRetry, err := c.httpClient.Do(reqRetry)
			if err != nil {
				return nil, err
			}
			defer respRetry.Body.Close()
			resp = respRetry
		}
	}

	if resp.StatusCode == http.StatusUnauthorized {
		return nil, fmt.Errorf("Transmission authentication failed, check credentials")
	}
	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("Transmission HTTP error %d: %s", resp.StatusCode, string(bodyBytes))
	}

	var res rpcResponse
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return nil, err
	}

	if res.Result != "success" {
		return nil, fmt.Errorf("Transmission RPC error: %s", res.Result)
	}

	return &res, nil
}

func (c *TransmissionClient) GetActiveTorrentCount(ctx context.Context) (int, error) {
	logger.Debug("<trans|%s> Getting active torrent count", c.clientConfig.URL)

	res, err := c.doRPC(ctx, "session-stats", nil)
	if err != nil {
		return 0, err
	}

	if countVal, ok := res.Arguments["activeTorrentCount"]; ok {
		if countFloat, ok := countVal.(float64); ok {
			return int(countFloat), nil
		}
	}
	return 0, nil
}

func (c *TransmissionClient) SetUploadSpeed(ctx context.Context, speed float64) error {
	var args map[string]interface{}
	if math.IsInf(speed, 1) {
		logger.Debug("<trans|%s> Setting upload speed to unlimited", c.clientConfig.URL)
		args = map[string]interface{}{
			"speed-limit-up-enabled": false,
		}
	} else {
		logger.Debug("<trans|%s> Setting upload speed to %v%s", c.clientConfig.URL, speed, c.appConfig.Units)
		converted, err := units.Convert(speed, c.appConfig.Units, "KB")
		if err != nil {
			return err
		}
		speedLimitKB := int64(math.Max(1, math.Round(converted)))
		args = map[string]interface{}{
			"speed-limit-up-enabled": true,
			"speed-limit-up":         speedLimitKB,
		}
	}

	_, err := c.doRPC(ctx, "session-set", args)
	return err
}

func (c *TransmissionClient) SetDownloadSpeed(ctx context.Context, speed float64) error {
	var args map[string]interface{}
	if math.IsInf(speed, 1) {
		logger.Debug("<trans|%s> Setting download speed to unlimited", c.clientConfig.URL)
		args = map[string]interface{}{
			"speed-limit-down-enabled": false,
		}
	} else {
		logger.Debug("<trans|%s> Setting download speed to %v%s", c.clientConfig.URL, speed, c.appConfig.Units)
		converted, err := units.Convert(speed, c.appConfig.Units, "KB")
		if err != nil {
			return err
		}
		speedLimitKB := int64(math.Max(1, math.Round(converted)))
		args = map[string]interface{}{
			"speed-limit-down-enabled": true,
			"speed-limit-down":         speedLimitKB,
		}
	}

	_, err := c.doRPC(ctx, "session-set", args)
	return err
}
